integration-tests/deployment/keystone/deploy.go

```diff
@@ -0,0 +1,537 @@
+			ids = append(ids, n.ID)
+		}
+		donToNodeIDs[donName] = ids
+	}
+
+	// register capabilities
+	capabilitiesResp, err := registerCapabilities(lggr, registerCapabilitiesRequest{
+		chain:                registryChain,
+		registry:             registry,
+		donToCapabilities:    req.DonToCapabilities,
+		donToOffchainNodeIDs: donToNodeIDs,
+	})
+	if err != nil {
+		return nil, fmt.Errorf("failed to register capabilities: %w", err)
+	}
+	lggr.Infow("registered capabilities", "capabilities", capabilitiesResp.donToCapabilities)
+
+	// register node operators
+	var nops []capabilities_registry.CapabilitiesRegistryNodeOperator
+	for _, nop := range req.NodeIDToNop {
+		nops = append(nops, nop)
+	}
+	nopsResp, err := registerNOPS(ctx, registerNOPSRequest{
+		chain:    registryChain,
+		registry: registry,
+		nops:     nops,
+	})
+	if err != nil {
+		return nil, fmt.Errorf("failed to register node operators: %w", err)
+	}
+	lggr.Infow("registered node operators", "nops", nopsResp.nops)
+
+	nopToNodeIDs := make(map[capabilities_registry.CapabilitiesRegistryNodeOperator][]string)
+	for nodeID, nop := range req.NodeIDToNop {
+		if _, ok := nopToNodeIDs[nop]; !ok {
+			nopToNodeIDs[nop] = make([]string, 0)
+		}
+		nopToNodeIDs[nop] = append(nopToNodeIDs[nop], nodeID)
+	}
+	nodeToRegisterNop := make(map[string]*capabilities_registry.CapabilitiesRegistryNodeOperatorAdded)
+	for _, nop := range nopsResp.nops {
+		n := capabilities_registry.CapabilitiesRegistryNodeOperator{
+			Name:  nop.Name,
+			Admin: nop.Admin,
+		}
+		nodeIDs := nopToNodeIDs[n]
+		for _, nodeID := range nodeIDs {
+			_, exists := nodeToRegisterNop[nodeID]
+			if !exists {
+				nodeToRegisterNop[nodeID] = nop
+			}
+		}
+	}
+
+	//	nopOnchainIDtoParams := make(map[uint32]capabilities_registry.CapabilitiesRegistryNodeParams)
+	nodeIDToParams := make(map[string]capabilities_registry.CapabilitiesRegistryNodeParams)
+	for don, ocr2nodes := range donToOcr2Nodes {
+		caps, ok := capabilitiesResp.donToCapabilities[don]
+		var hashedCapabilityIds [][32]byte
+		for _, cap := range caps {
+			hashedCapabilityIds = append(hashedCapabilityIds, cap.id)
+		}
+		lggr.Debugw("hashed capability ids", "don", don, "ids", hashedCapabilityIds)
+		if !ok {
+			return nil, fmt.Errorf("capabilities not found for node operator %s", don)
+		}
+		for _, n := range ocr2nodes {
+			if n.IsBoostrap { // bootstraps are part of the DON but don't host capabilities
+				continue
+			}
+			nop, ok := nodeToRegisterNop[n.ID]
+			if !ok {
+				return nil, fmt.Errorf("node operator not found for node %s", n.ID)
+			}
+			params, ok := nodeIDToParams[n.ID]
+
+			if !ok {
+				params = capabilities_registry.CapabilitiesRegistryNodeParams{
+					NodeOperatorId:      nop.NodeOperatorId,
+					Signer:              n.Signer,
+					P2pId:               n.P2PKey,
+					HashedCapabilityIds: hashedCapabilityIds,
+				}
+			} else {
+				// when we have a node operator, we need to dedup capabilities against the existing ones
```

Comment:
I don't understand - why would we have duplicates? Is this to make the tool idempotent? But we're deploying a new Registry contract on every call...